src/commands/txCommands/pullAll.js

```diff
@@ -0,0 +1,86 @@
+const txlib = require('./util');
+const pullResourceTrns = require('./util/pullResourceTrns');
+
+const getProjectResourcesList$ =
+	txlib.resources$
+		.flatMap(Rx.Observable.from);
+
+const child_process$ = Rx.Observable.bindNodeCallback(child_process.exec);
+
+/**
+ * Creates git commit of what has been staged
+ *
+ * note: adjusting resource name in git commit message to prevent jira ticket
+ * from moving through workflow. We have a script which looks at all git commits
+ * to see if a ticket number is contained within it and if so associates a branch/PR
+ * with it. We don't want that to happen with tx pulls.
+ * @param  {String} resource resource slug to be used when we pull translations
+ * @return {Observable} Observable of shell process
+ */
+const commitResource$ = (resource) => {
+	const commitCommand = `git commit -m "tx:pull for ${resource.replace(/-/g, '_')}" --no-verify`;
+	return child_process$(commitCommand)
+		.do(() => {
+			console.log(chalk.green(commitCommand));
+		});
+};
+
+/**
+ * Called when a tx pull is complete for a particular resource
+ * @param  {String} resource resource slug to be used when we pull translations
+ * @param  {Boolean} commitOnComplete flag used to determine if we should commit the resource in git
+ * @return {Observable} Observable of tx process
+ */
+const onPullResourceComplete = (resource, commitOnComplete) => {
+	if (!commitOnComplete) {
+		return Rx.Observable.empty();
+	}
+
+	return child_process$('git status')
+		.flatMap(([stdout, stderr]) => {
+			if (!stdout.includes('modified:')) {
+				console.log(chalk.grey(`no changes to commit for '${resource}'`));
+				return Rx.Observable.empty();
+			}
+
+			child_process.execSync('git add src/trns');
+			return commitResource$(resource);
+		});
+};
+
+/**
+ * Kicks off process to pull an individual resources trns
+ * @param  {String} resource resource slug to be used when we pull translations
+ * @param  {Boolean} commitOnComplete flag used to determine if we should commit the resource in git
+ * @return {Observable} Observable of tx process
+ */
+const pullResource = (resource, commitOnComplete) => {
+	console.log(chalk.cyan(`Starting tx:pull for '${resource}'`));
+	return pullResourceTrns.pullResourceContent$(resource)
+		.toArray() // wait until all content has been pulled before moving on to next tasks
+		.do(() => console.log(chalk.green(`\nCompleted tx:pull for '${resource}'`)))
+		.flatMap(() => onPullResourceComplete(resource, commitOnComplete));
+};
+
+module.exports = {
+	command: 'pullAll',
+	description: 'Pull all translations from resources from Transifex, ordered by most recently updated',
+	builder: yarg =>
+		yarg.option({
+			gitCommit: {
+				alias: 'c',
+				default: false
+			},
+		}),
+	handler: argv => {
+		txlib.checkEnvVars();
+		console.log(chalk.magenta('Start pulling all resources process...'));
+
+		getProjectResourcesList$
+			.flatMap(resource => pullResource(resource, argv.gitCommit), 1)
```

Comment:
i'm going to be annoying and recommend that you move the 'commitOnComplete' logic into this function rather than folding it into `pullResource` - I think it's better to cleanly separate 'pulling' from 'committing'

Essentially, that just means moving the `.flatMap` on line 65 to here and removing the `commitOnComplete` arg from `pullResource`. You can also then rename `onPullResourceComplete` to `commit$` and treat it as a generic 'committing' function that could be run for anything - you might even want to move it to `/txCommands/util/index.js` as a generic library function.

In that case, I think the `if (commitOnComplete)` check would also be specified here rather than in the `commit$` function

Reply:
👍  i'll get back to this this afternoon

Reply:
@mmcgahan I've tried moving `flatMap` as suggested, but the Observable that is returned doesn't have the `resource` slug which i need for the next steps. Is there something I need to return for the next step in flatMap?

Reply:
just call `onPullResourceComplete` while `resource` is in scope

```js
.flatMap(
  resource =>
    pullResource(resource, argv.gitCommit), 1)
      .flatMap(() => {
        if (!commitOnComplete) {
          return Observable.empty();
        }
        return onPullResourceComplete(resource)
      })
)
.subscribe(...)
```

Alternatively you could add `.map(() => resource)` to the end of `pullResource` so that it returns that value when it completes (as sort of a 'success' signal)

then it would just be

```js
.flatMap(resource => pullResource(resource, argv.gitCommit), 1))
.flatMap(resource => {
  if (!commitOnComplete) {
    return Observable.empty();
  }
  return onPullResourceComplete(resource)
})
.subscribe(...)
```